SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriverRPC.cpp

```diff
@@ -0,0 +1,451 @@
+			}
+			OutUpdates.Last().Spans.Add(NewSpanId);
+		}
+	};
+}
+
+RPCCallbacks::DataWritten USpatialNetDriverRPC::MakeDataWriteCallback(TArray<FWorkerComponentData>& OutArray) const
+{
+	return [&OutArray](Worker_EntityId EntityId, Worker_ComponentId ComponentId, Schema_ComponentData* InData) {
+		if (ensure(InData != nullptr))
+		{
+			FWorkerComponentData Data;
+			Data.component_id = ComponentId;
+			Data.schema_type = InData;
+			OutArray.Add(Data);
+		}
+	};
+}
+
+RPCCallbacks::UpdateWritten USpatialNetDriverRPC::MakeUpdateWriteCallback(TArray<UpdateToSend>& OutUpdates) const
+{
+	return [&OutUpdates](Worker_EntityId EntityId, Worker_ComponentId ComponentId, Schema_ComponentUpdate* InUpdate) {
+		if (ensure(InUpdate != nullptr))
+		{
+			if (OutUpdates.Num() == 0 || OutUpdates.Last().EntityId != EntityId || OutUpdates.Last().Update.component_id != ComponentId)
+			{
+				OutUpdates.AddDefaulted();
+				OutUpdates.Last().EntityId = EntityId;
+				OutUpdates.Last().Update.component_id = ComponentId;
+			}
+			OutUpdates.Last().Update.schema_type = InUpdate;
+		}
+	};
+}
+
+void USpatialNetDriverRPC::Init(USpatialNetDriver* InNetDriver, const SpatialGDK::FSubView& InActorAuthSubView,
+								const SpatialGDK::FSubView& InActorNonAuthSubView)
+{
+	NetDriver = InNetDriver;
+	LatencyTracer = USpatialLatencyTracer::GetTracer(InNetDriver->GetWorld());
+	EventTracer = InNetDriver->Connection->GetEventTracer();
+	RPCService = MakeUnique<SpatialGDK::RPCService>(InActorNonAuthSubView, InActorAuthSubView);
+
+	{
+		// TODO
+		// MulticastReceiver =
+	}
+}
+
+void USpatialNetDriverRPC::AdvanceView()
+{
+	RPCService->AdvanceView();
+}
+
+void USpatialNetDriverRPC::ProcessReceivedRPCs()
+{
+	// MulticastReceiver->
+}
+
+TArray<FWorkerComponentData> USpatialNetDriverRPC::GetRPCComponentsOnEntityCreation(const Worker_EntityId EntityId)
+{
+	static TArray<Worker_ComponentId> EndpointComponentIds = {
+		SpatialConstants::CLIENT_ENDPOINT_COMPONENT_ID, SpatialConstants::SERVER_ENDPOINT_COMPONENT_ID,
+		/*SpatialConstants::MULTICAST_RPCS_COMPONENT_ID,
+		SpatialConstants::CROSSSERVER_SENDER_ENDPOINT_COMPONENT_ID*/
+	};
+
+	TArray<FWorkerComponentData> Components;
+	GetRPCComponentsOnEntityCreation(EntityId, Components);
+
+	for (auto EndpointId : EndpointComponentIds)
+	{
+		if (Components.FindByPredicate([EndpointId](const FWorkerComponentData& Data) {
+				return Data.component_id == EndpointId;
+			})
+			== nullptr)
+		{
+			FWorkerComponentData Data;
+			Data.component_id = EndpointId;
+			Data.schema_type = Schema_CreateComponentData();
+			Components.Add(Data);
+		}
+	}
+
+	return Components;
+}
+
+void USpatialNetDriverRPC::GetRPCComponentsOnEntityCreation(const Worker_EntityId EntityId, TArray<FWorkerComponentData>& OutData) {}
+
+void USpatialNetDriverRPC::FlushRPCUpdates()
+{
+	TMap<FName, RPCService::RPCReceiverDescription>& Receivers = RPCService->GetReceivers();
+	TArray<UpdateToSend> Updates;
+	for (auto Receiver : Receivers)
```

Comment:
```suggestion
	for (auto& Receiver : Receivers)
```

Reply:
Yeah, const would work, it's just a description structure.